### ISSUE TITLE
New rule suggestion: Implementations of Equals(T) for IEquatable<T> should be virtual for non-sealed classes

### ISSUE THREAD

**Issue by matteo-prosperi** [dotnet/runtime#52957]
According to the documentation of `IEquatable`, "If you implement IEquatable, you should also override the base class implementations of Equals(Object) and GetHashCode() so that their behavior is consistent with that of the Equals(T) method". In my opinion this implies that Equals(T) and Equals(Object) behaves in the same way.
Because `Equals(Object)` is a virtual method, `Equals(T)` should also be required to be virtual (unless the type is sealed) in order to allow inheritors to maintain coherence between the two implementations.

### Describe suggestions on how to achieve the rule

Any time `IEquatable<T>` is implemented for a non-sealed class, the `Equals(T)` method should be required to be virtual.

**Comment by FilipToth** [comment]
Yes, I suppose we could throw an analyzer error in that case, throwing an exception would be completely unnecessary in my opinion.

**Comment by DaZombieKiller** [comment]
The analyzer should also consider the `sealed override` case:
```cs
class MyEquatableClass : IEquatable<MyEquatableClass>
{
    public int Value { get; }

    public MyEquatableClass(int value)
    {
        Value = value;
    }

    public sealed override int GetHashCode()
    {
        return Value;
    }

    public sealed override bool Equals(object obj)
    {
        return obj is MyEquatableClass other && Equals(other);
    }

    public bool Equals(MyEquatableClass other)
    {
        return Value == other.Value;
    }
}
```

**Comment by tannergooding** [comment]
Closing given the age and relatively minor amount of interest shown.

Every analyzer added has potentially large cost to all builds. It is relatively simple to roll your own analyzers if you need that for your code.

While the implementations are tied together, there are various special considerations that can come into play when trying to equate unsealed types, particularly against a concrete type, so its a much broader design issue to be considered and the impact that will occur for `IEquatable<SpecificType>`
